### PR TITLE
Add the ability to set the default theme through a configuration file

### LINKED ISSUE #13: Add the ability to be able to set a default theme
It would be cool to be able to specify a default theme somehow. Maybe in a config file like `~/.config/chime.conf`?

```ini
[chime]
theme = zelda
```

### PR DESCRIPTION
**Pull request by paw-lu**
This PR adds the ability to set the default theme via a configuration file.

The location is set on `~/.config/chime/chime.conf` on Unix or `%APPDATA%\chime\chime.ini` on Windows.

The format of the configuration file is:

```ini
[chime]
theme = zelda

```

Two unit tests have been added that test the detection of the config file location based on platform and test that chime changes its default theme based on the config file.

Closes #13

### CODE REVIEW

File: test_chime.py [hunk 1]
```diff
@@ -35,3 +39,26 @@ def test_script():
...
+    monkeypatch.setenv('APPDATA', '/Users/chime/AppData/Roaming')
+    config_path = chime._get_config_path(system)
+    assert config_path.as_posix() == expected_config_path
+
+
+def test__get_default_theme():
+    config_text = textwrap.dedent('''\
+    [chime]
+    theme = zelda
+    ''')
```
Comment (MaxHalford) [inline test_chime.py]: I would just add a `assert chime.theme() == 'chime'` to make it obvious that the theme has changed :)

File: chime.py [hunk 2]
```diff
@@ -17,7 +19,48 @@
...
+
+    Parameters:
+        system: The OS being used.
+    """
+    home_dir = pathlib.Path.home()
+    if system == "Windows":
+        config_path = (
+            pathlib.Path(
+                         os.getenv('APPDATA',
+                                   home_dir / pathlib.Path("AppData", "Roaming")))
```
Comment (MaxHalford) [inline chime.py]: Can you use single quotes instead of double quotes? I've forgotten what's PEP8 but I guess we should align.

**Comment by MaxHalford** [comment]
The PR is really cool, I didn't know about pytest monkeypatching!

File: test_chime.py (hunk 1 above)
Comment (paw-lu) [reply test_chime.py]: You pointing this out made me realize there was some fragility around this whole test design 😅, so I ended up refactoring it, but followed your general advice—asserting `chime.theme()` twice to check for a difference.

File: chime.py (hunk 2 above)
Comment (paw-lu) [reply chime.py]: Good catch! Tried to match your code style as much as possible but some things slipped. Feel free to let me know if other things look off!

File: test_chime.py [hunk 3]
```diff
@@ -1,12 +1,30 @@
...
 import pytest
 
 import chime
 
 
+@pytest.fixture(scope='function', autouse=True)
+def reload_chime():
+    importlib.reload(chime)
+
+
```
Comment (paw-lu) [inline test_chime.py]: So there was some fragility around testing the theme.

First of all, if one test modifies `THEME`—like `test__get_config_path` currently does—this could end up altering the state of some tests. So if a test ends on chaining the theme to `"zelda"`, and the next test ran `chime.theme()`, it would return `"zelda"`.

`reload_chime` just reloads chime before every test to reset the state in an attempt to fix this.

File: test_chime.py [hunk 4]
```diff
@@ -25,13 +43,47 @@ def test_no_exception():
...
+    config_path = chime._get_config_path(system)
+    assert config_path == expected_config_path
+
+
+def test_config_file(monkeypatch: _pytest.monkeypatch.MonkeyPatch):
+    assert chime.theme() == 'chime'
+    config_text = textwrap.dedent("""\
+    [chime]
+    theme = zelda
+    """)
```
Comment (paw-lu) [inline test_chime.py:67]: `"""` or `'''` here?

File: test_chime.py [hunk 5]
```diff
@@ -1,12 +1,31 @@
...
+
+
+@pytest.fixture(scope='function', autouse=True)
+def mock_pathlib_home(monkeypatch: _pytest.monkeypatch.MonkeyPatch):
+    with tempfile.TemporaryDirectory() as home_dir:
+        home_dir_path = pathlib.Path(home_dir)
+        monkeypatch.setattr(pathlib.Path, name='home', value=lambda: home_dir_path)
+        monkeypatch.setenv('APPDATA', value=str(pathlib.Path(home_dir)))
+
+
```
Comment (paw-lu) [inline test_chime.py]: Another problem can happen if the user running the test has a config file on their machine. Their config could specify another default theme, changing the theme the test run on.

`mock_pathlib_home` attempts to fix this by making a temporary directory and having it take over as the home directory (or `APPDATA` directory on Windows), where there will be no config file and chime will use the standard default theme across all tests until it is purposely changed.

**Comment by paw-lu** [comment]
There's one problem left to solve. Pytest (to my surprise) automatically picked up on `:81` of `README.md` and treats it as a doctest.

```md
```py
>>> chime.theme()  # return the current theme
'chime'

```

If the user has a config file that changes the default theme this test fill fail.

File: test_chime.py (hunk 4 above)
Comment (MaxHalford) [reply test_chime.py:67]: `"""` is fine :)

File: test_chime.py (hunk 5 above)
Comment (MaxHalford) [reply test_chime.py]: Very smart!

**Comment by MaxHalford** [comment]
Yes I'm not sure if pytest has a global setup function to mock the config file before running tests, in any case I can't seem to find such a thing. I would say that things are fine like this, it's a bit of an edge case. However, the codebase quality of chime is currently quite high, and we're not in a rush. I don't have the time to look into this right now (work + hackathon this weekend), but let's take our time :)

Edit: maybe you can this via the `conftest.py` file, as mentioned in [this SO thread](https://stackoverflow.com/questions/26899001/pytest-how-do-i-use-global-session-wide-fixtures).

**Comment by paw-lu** [comment]
> maybe you can this via the conftest.py file

Yeah that did the trick. `README.md` now passes even if the user has a config file changing the default theme.

> I don't have the time to look into this right now

Take your time! Never any rush on any of these 😀.

**Comment by MaxHalford** [comment]
It's weird, I don't see the builds being built :/

**Comment by paw-lu** [comment]
That's a surprise for me too. [Everything passed on my branch](https://github.com/paw-lu/chime/runs/1432278509), but I'm not sure why things aren't kicking off here.

**Comment by paw-lu** [comment]
#16 hopefully fixes this

**Comment by MaxHalford** [comment]
I guess you need to bump this pull request to trigger the CI :)

**Comment by paw-lu** [comment]
Rebased onto main!

**Comment by MaxHalford** [comment]
Released in version `0.6.0` :). Thanks again!
